### PR TITLE
Cleanup and finish the definition of various script related classes

### PR DESCRIPTION
**Pull request by jonwil**

### CODE REVIEW

File: src/game/common/rts/buildinfo.cpp [hunk 1]
```diff
@@ -127,20 +127,22 @@ BuildListInfo &BuildListInfo::operator=(const BuildListInfo &that)
...
-    m_location.z = input.Read_Real32();
-    m_angle = input.Read_Real32();
-    m_isInitiallyBuilt = input.Read_Byte() != 0;
-    m_numRebuilds = input.Read_Int32();
+    Set_Building_Name(input.Read_AsciiString());
+    Set_Template_Name(input.Read_AsciiString());
+    Coord3D location;
+    location.x = input.Read_Real32();
+    location.y = input.Read_Real32();
+    location.z = 0; // Confirmed that this throws away the return value of Read_Real
```
Comment (xezon) [inline src/game/common/rts/buildinfo.cpp:136]: Where is the matching "Write_Data_Chunk" function for this? If there is one?

File: src/game/common/rts/buildinfo.h [hunk 2]
```diff
@@ -47,17 +47,74 @@ class BuildListInfo : public MemoryPoolObject, public SnapShot
...
+    bool Get_Whiner() const { return m_whiner; }
+    bool Get_Repairable() const { return m_repairable; }
+    RenderObjClass *Get_Render_Obj() const { return m_renderObj; }
+    ObjectID Get_Object_ID() const { return m_objectID; }
+    int Get_Max_Resource_Gatherers() const { return m_maxResourceGatherers; }
+    int Get_Resource_Gatherers() const { return m_resourceGatherers; }
+    const Coord2D *Get_Rally_Point_Offset() const { return &m_rallyPointOffset; }
+    unsigned int Get_Object_Timestamp() const { return m_objectTimestamp; }
+    bool Get_Unk4() const { return m_unkbool4; }
+    bool GeT_Auto_Build() const { return m_autoBuild; }
```
Comment (xezon) [inline src/game/common/rts/buildinfo.h]: Get_Auto_Build

File: src/game/common/rts/sideslist.h [hunk 3]
```diff
@@ -44,28 +44,37 @@ class SidesList : public SubsystemInterface, public SnapShot
...
+    bool Validate_Ally_Enemy_List(const Utf8String &team, Utf8String &allies);
+    void Add_Player_By_Template(Utf8String template_name);
+    SidesInfo *Find_Side_Info(Utf8String name, int *index = nullptr);
+    SidesInfo *Find_Skirmish_Side_Info(Utf8String name, int *index = nullptr);
+    SidesInfo *Get_Side_Info(int index);
+    SidesInfo *Get_Skirmish_Side_Info(int index);
     void Add_Side(const Dict *dict);
+    void Remove_Side(int index);
+    void Prepare_For_MP_Or_Skirmish();
+    bool Is_Player_Default_Team(TeamsInfo *t);
```
Comment (xezon) [inline src/game/common/rts/sideslist.h:59]: can be const?

File: src/game/common/rts/team.h [hunk 4]
```diff
@@ -341,10 +341,13 @@ class Team : public MemoryPoolObject, public SnapShot
     const TeamPrototype *Get_Prototype() const { return m_proto; }
     const Waypoint *Get_Current_Waypoint() { return m_currentWaypoint; }
     bool Did_Enter_Or_Exit() { return m_enteredOrExited; }
+    Utf8String &Get_State() { return m_state; }
+    bool Get_Created() { return m_created; }
```
Comment (xezon) [inline src/game/common/rts/team.h]: const

File: src/game/logic/scriptengine/scriptcondition.h [hunk 5]
```diff
@@ -154,8 +156,31 @@ class Condition : public MemoryPoolObject
...
+    {
+        if (ndx < 0 || ndx >= m_numParams) {
+            return nullptr;
+        } else {
+            return m_params[ndx];
+        }
+    }
 
-    static bool Parse_Data_Chunk(DataChunkInput &input, DataChunkInfo *info, void *data);
+    void Set_Warinings(bool set) { m_hasWarnings = set; }
```
Comment (xezon) [inline src/game/logic/scriptengine/scriptcondition.h]: Set_Warnings

File: src/game/logic/scriptengine/scriptcondition.h [hunk 6]
```diff
@@ -179,10 +204,144 @@ class OrCondition : public MemoryPoolObject
...
+    virtual bool Evaluate_Team_Is_Contained(Parameter *param, bool b) = 0;
+};
+
+class ScriptConditions : public ScriptConditionsInterface
+{
+public:
+    ScriptConditions();
+    virtual ~ScriptConditions() override;
+    virtual void Init() override;
+    virtual void PostProcessLoad() {}
```
Comment (xezon) [inline src/game/logic/scriptengine/scriptcondition.h]: override

File: src/game/logic/scriptengine/scriptengine.h [hunk 7]
```diff
@@ -177,19 +178,94 @@ class ScriptEngine : public SubsystemInterface, public SnapShot
...
     bool Is_Time_Frozen_Script() { return m_freezeByScript; }
     bool Is_Time_Frozen_Debug();
     bool Is_Time_Fast();
-    void Notify_Of_Team_Destruction(Team *team_destroyed);
     void Append_Debug_Message(const Utf8String &message, bool b);
     GameDifficulty Get_Difficulty() { return m_gameDifficulty; }
     const AttackPriorityInfo *Get_Attack_Info(Utf8String const &name);
+    bool Has_Shown_MP_Local_Defeat_Window() const { return m_hasShowMPLocalDefeatWindow; }
+    void Mark_MP_Local_Defeat_Window_Shown() { m_hasShowMPLocalDefeatWindow = true; }
+    bool Is_End_Game_Timer_Running() { return m_endGameTimer >= 0; }
```
Comment (xezon) [inline src/game/logic/scriptengine/scriptengine.h]: const

File: src/game/common/rts/buildinfo.h (hunk 2 above)
Comment (jonwil) [reply src/game/common/rts/buildinfo.h]: fixed.

File: src/game/logic/scriptengine/scriptcondition.h (hunk 5 above)
Comment (jonwil) [reply src/game/logic/scriptengine/scriptcondition.h]: fixed.

File: src/game/logic/scriptengine/scriptcondition.h (hunk 6 above)
Comment (jonwil) [reply src/game/logic/scriptengine/scriptcondition.h]: Fixed.

File: src/game/common/rts/buildinfo.cpp (hunk 1 above)
Comment (jonwil) [reply src/game/common/rts/buildinfo.cpp:136]: There isn't one, it's done inside SidesList::Write_Sides_Data_Chunk

File: src/game/common/rts/team.h (hunk 4 above)
Comment (jonwil) [reply src/game/common/rts/team.h]: fixed.

File: src/game/logic/scriptengine/scriptengine.h (hunk 7 above)
Comment (jonwil) [reply src/game/logic/scriptengine/scriptengine.h]: fixed.

File: src/game/common/rts/buildinfo.cpp (hunk 1 above)
Comment (xezon) [reply src/game/common/rts/buildinfo.cpp:136]: And the written bytes match the read bytes here? Aka just 8 bytes for location?

Comment (jonwil) [reply src/game/common/rts/buildinfo.cpp:136]: Another check shows that it does make a 3rd call to Read_Real but ignores the result. Fixed, it matches what Write does.

**Review by xezon — APPROVED** [review]
Tested Shell map and USA 1. No crashes.

File: src/game/common/rts/buildinfo.cpp [hunk 8]
```diff
@@ -25,18 +25,18 @@ BuildListInfo::BuildListInfo() :
     m_script(),
     m_health(100),
     m_whiner(true),
-    m_repairable(false),
-    m_sellable(true),
+    m_sellable(false),
+    m_repairable(true),
```
Comment (bobtista) [inline src/game/common/rts/buildinfo.cpp:29]: Curious, why have we flipped the boolean values here for m_sellable and m_repairable?

Comment (xezon) [reply src/game/common/rts/buildinfo.cpp:29]: @jonwil

Comment (xezon) [reply src/game/common/rts/buildinfo.cpp:29]: Jonwil says this is correct

**Pull request merged**